### ISSUE TITLE
Regarding training process and best.pt model

### ISSUE THREAD

**Issue by UTKARSHS1994** [ultralytics/ultralytics#3972]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
Thanks for this wonderful work. This is awesome.

I am training a model for 100 epochs. I have 2 questions which I need to clarify. Please see the below screenshot of the last epochs of training
![image](https://github.com/ultralytics/ultralytics/assets/54258787/388f1b66-668b-4be0-beed-e59c1db39af3)

1) I believe that the metrics reported at the end  of every epoch (shown in red) is model performance on **training** dataset while the metrics reported at the end of the training (shown in blue) is model performance on **validation** dataset. Is this correct? If not, then please tell what the metrics shown in red and blue actually represent?

2) Also, when we use the model "best.pt", what is it best in terms of and on which dataset? I believe it is best in terms of having highest mAP on training dataset.

Thanks a lot.

**Comment by glenn-jocher** [comment]
@UTKARSHS1994 hello,

I'm happy to hear that you're finding value in our work with YOLOv8!

You're partially correct in your interpretations of the color-coding in the terminal output.

1) The metrics highlighted in red at the end of each epoch signify the model performance on the training dataset for that given epoch. This is correct. However, the metrics showcased in blue at the completion of training represent the performance of the 'best' model on the validation dataset. This 'best' model is selected during the course of the entire training based on your criterion of choice (by default, highest mAP).

2) When it comes to the 'best.pt' model, it is selected on the basis of the performance during validation, not training. As its name suggests, 'best.pt' is the model that achieved the 'best' evaluation metric (by default, the highest mean Average Precision, or mAP) on the validation dataset at any point during the entire training period. 

Therefore, the 'best.pt' model doesn't necessarily correlate with the final epochs of training; it could be from any epoch wherein the model achieved the highest validation performance.

I hope this clarifies your questions! Let me know if there's anything else you need assistance with.

Best,
Glenn

**Comment by UTKARSHS1994** [comment]
@glenn-jocher 
Thanks for your reply!

1) I think I understand it now. So after every epoch in training, the model's performance is checked on both **training** and **validation** datasets, although only training results are displayed (the metrics highlighted in red). The validation results are not displayed after every epoch, only the 'best' validation results are displayed at the end of training. Correct?

2) You said that the best model is selected based on my criterion of choice and it is **mAP** by default. I think you mean mAP50 (and not mAP50-95). However in a different issue of last month, I checked that the default metric is precision. [](https://github.com/ultralytics/ultralytics/issues/3435).
Please confirm which is the default metric and do we have the choice of metric as you suggested? (because the choice was not available a month back)

Thanks a lot!

**Comment by glenn-jocher** [comment]
Hello @UTKARSHS1994,

I appreciate your follow-up questions.

1) Correct! After every epoch in training, the model's performance is assessed on both the training and validation datasets. However, for the sake of simplicity and clarity in the console output, only the training results are immediately displayed for each epoch (the metrics shown in red). The best validation results—those produced by the 'best' model—are then displayed at the end of the entire training process.

2) Concerning the selection metric for the 'best' model, as of the current version of YOLOv8, yes, the default metric is indeed mAP50. It's possible that the metric was precision in a previous version or in a different context. However, as per the latest state, mAP50 is the default criterion. 

Moreover, it is indeed possible to switch the criterion to one that best suits your particular application or requirements. This feature provides flexibility in how the 'best' model is determined based on your project's unique demands.

I hope this further clarifies your concerns. Happy training, and don't hesitate to return if you have more questions!

Best,
Glenn

**Comment by UTKARSHS1994** [comment]
Hello  @glenn-jocher,

Thanks for your wonderful replies.

How to switch the metric from mAP50 to say, mAP50-95? 

I didn't find any argument in the args.yaml file for the same. Am I missing something?

Thanks

**Comment by glenn-jocher** [comment]
Hello @UTKARSHS1994,

As you've noticed, there isn't a direct command line argument or a parameter in args.yaml to switch the metric from mAP50 to mAP50-95.

However, behind the scenes, the model does calculate various performance metrics including mAP at multiple Intersection Over Union (IOU) thresholds i.e., .50, .55, ..., .95 during validation but uses mAP50 as default to select the best model.

Currently, there's no out-of-the-box parameter to make the 'best' model determination based on mAP50-95. It would require modifications to the source code to change the metric which determines the 'best' model.

I hope this clears the ambiguity. Should you decide to go forward with such modifications and run into any issues, feel free to return and I'll be happy to assist.

Best,
Glenn

**Comment by UTKARSHS1994** [comment]
Hello @glenn-jocher, 

Thanks a lot for your clear and concise replies.

**Comment by glenn-jocher** [comment]
Hello @UTKARSHS1994,

You're most welcome! I'm glad I could provide you with clear and concise responses to your queries. That's what we strive for in developing Ultralytics and the YOLOv8 repo - to create not only highly efficient and accurate models, but also to ensure they're user-friendly and well-documented.

Feel free to come back with any other questions or issues you may have regarding YOLOv8. The community and I are here to help.

Best,
Glenn

**Comment by harakas** [comment]
I think the answer to #2 can be found in https://github.com/ultralytics/yolov5/issues/582 -- best.pt is selected based on a weighted sum of mAP50 and mAP50-90. Ignore the glenn-bot/chatgpt hallucination.

See the current codebase:
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py#L742-L745

```python
    def fitness(self):
        """Model fitness as a weighted combination of metrics."""
        w = [0.0, 0.0, 0.1, 0.9]  # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
        return (np.array(self.mean_results()) * w).sum()
```

**Comment by glenn-jocher** [comment]
@harakas thanks for pointing that out! 🌟 You're right, the selection of `best.pt` indeed involves a weighted sum of mAP50 and mAP50-95 metrics. My earlier response might have oversimplified things a bit - appreciate the correction!

For those curious, here's a quick peek at the relevant code snippet for clarity:
```python
# Weighted sum of mAP50 and mAP50-95
best_metric = mAP50 * 0.9 + mAP50-95 * 0.1
```
Thanks again for keeping the information accurate and up-to-date!

**Comment by CsterKuroi** [comment]
> @harakas thanks for pointing that out! 🌟 You're right, the selection of `best.pt` indeed involves a weighted sum of mAP50 and mAP50-95 metrics. My earlier response might have oversimplified things a bit - appreciate the correction!
> 
> For those curious, here's a quick peek at the relevant code snippet for clarity:
> 
> ```python
> # Weighted sum of mAP50 and mAP50-95
> best_metric = mAP50 * 0.9 + mAP50-95 * 0.1
> ```
> 
> Thanks again for keeping the information accurate and up-to-date!

best_metric =  mAP50-95 * 0.9 + mAP50 * 0.1 ?

**Comment by MAK1647** [comment]
Is that posible if we want to point out the .pt which have highest mAP50 ? If yes how to configure it?
I ask this question because in reseaching, sometime i need to compare some metrics with others published article. And in that case we need to point out which .pt is the best base on specifically metric (such as mAP90 or Precesion).

**Comment by sydjovi** [comment]
how do i check the values in best.pt? How I view it?

**Comment by glenn-jocher** [comment]
You can check the values in `best.pt` by loading the model in Python using PyTorch and then inspecting its attributes. Here's a quick example:

```python
import torch

# Load the model
model = torch.load('

**Comment by MAK1647** [comment]
> Yes, it is possible to select the model based on a specific metric like mAP50. However, this requires modifying the source code to change the selection criterion. Currently, YOLOv8 uses a weighted sum of mAP50 and mAP50-95 by default. To customize this, you would need to adjust the logic in the training script to prioritize your desired metric. If you need further assistance with this, feel free to ask.

Hello @pderrenger , 
Thank you for your reply.

Currently, i would like to change the best.pt metric to **only mAP@50.**
I tried several times before. However, it still not work. 
Could you give me some assistances for this work? Which file or code line could be changed ?

**Comment by Y-T-G** [comment]
@MAK1647 You have to change the fitness weights. Making the weight for mAP50 1 and the rest 0.

https://github.com/ultralytics/ultralytics/blob/92cc8b8b52fc7122848d2d5472b79841135f902d/ultralytics/utils/metrics.py#L745-L748

**Comment by MAK1647** [comment]
> @MAK1647 You have to change the fitness weights. Making the weight for mAP50 1 and the rest 0.
> 
> https://github.com/ultralytics/ultralytics/blob/92cc8b8b52fc7122848d2d5472b79841135f902d/ultralytics/utils/metrics.py#L745-L748

Thank you!

**Comment by FlorianMinguet** [comment]
> @harakas thanks for pointing that out! 🌟 You're right, the selection of `best.pt` indeed involves a weighted sum of mAP50 and mAP50-95 metrics. My earlier response might have oversimplified things a bit - appreciate the correction!
> 
> For those curious, here's a quick peek at the relevant code snippet for clarity:
> 
> ```python
> # Weighted sum of mAP50 and mAP50-95
> best_metric = mAP50 * 0.9 + mAP50-95 * 0.1
> ```
> 
> Thanks again for keeping the information accurate and up-to-date!

Hi @glenn-jocher, first of all, thank you for all the detailed explanations !

I have two questions:
- Is the formula "best_metric = mAP50 * 0.9 + mAP50-95 * 0.1" the complete criterion formula, and is it the only factor used to select the best model weights?
- How can the "best" model have high values for the mAP50 and mAP50-95 indicators but very low values for the box(P) metric?

Here’s the context: 
I just trained a model with only one class, and the selected "best" model has the following metrics:
mAP50 = 0.958, mAP50-95 = 0.84.
These values seem fine, but the box(P) metric is very low at 0.00333. This seems strange because obviously when I use this "best" model, it doesn't make any predictions at all. It doesn't even attempt to detect anything, as if no predictions are being made. However, when I use the "last.pt" model, predictions work fine. I assume "last.pt" corresponds to the most recent weights from the validation dataset.

I loaded both "best.pt" and "last.pt" models using Pytorch and here are the "train_metrics" for both models, which I assume correspond to the metrics of the selected weights during the validation step based on the fitness function : 

-  "best.pt" : 
```'train_metrics': {'metrics/precision(B)': 0.00333, 'metrics/recall(B)': 1.0, 'metrics/mAP50(B)': 0.95815, 'metrics/mAP50-95(B)': 0.84039, 'val/box_loss': 0.42948, 'val/cls_loss': 2.99384, 'val/dfl_loss': 1.1884, 'fitness': 0.85217}```
- "last.pt" : 
```'train_metrics': {'metrics/precision(B)': 0.9961, 'metrics/recall(B)': 1.0, 'metrics/mAP50(B)': 0.995, 'metrics/mAP50-95(B)': 0.82756, 'val/box_loss': 0.66462, 'val/cls_loss': 0.65847, 'val/dfl_loss': 1.44658, 'fitness': 0.8443}```

Looking at these metrics, we see that the fitness value for "best.pt" (0.85217) is better than for "last.pt" (0.8443). But if we use the formula for calculating the best_metric, which is:

```best_metric = mAP50 * 0.9 + mAP50-95 * 0.1```

Here’s how it works out:

"best.pt":
0.95815 * 0.9 + 0.84039 * 0.1 = 0.946374 (which is different from the reported fitness value of 0.85217)

"last.pt":
0.995 * 0.9 + 0.82756 * 0.1 = 0.978256 (which is different from the reported fitness value of 0.8443)

So, based on the formula, "last.pt" appears to be superior (0.978 > 0.946) and the fitness values don't match what the formula suggests. It feels a bit messy, and I would appreciate it if you could clarify this discrepancy.

Thanks again!

**Comment by Y-T-G** [comment]
Did you try running model.val() with it after loading? Maybe it corrupted on save.

**Comment by Y-T-G** [comment]
Is this last.pt and best.pt after the training fully completed? Or did you interrupt the training?

**Comment by FlorianMinguet** [comment]
> Did you try running model.val() with it after loading? Maybe it corrupted on save.

No I didn't but I trained multiples times and I always ended up with the same results

**Comment by FlorianMinguet** [comment]
> Is this last.pt and best.pt after the training fully completed? Or did you interrupt the training?

Both were generated at the same time using the ```model.train(<parameters>)``` function and the training fully completed, I didn't interrupt the training

**Comment by Y-T-G** [comment]
The calculation is also the other way around.

`fitness = 0.9 * mAP50-95 + 0.1 * mAP50`

**Comment by Y-T-G** [comment]
How many images do you have? It seems like you have very few validation images.

**Comment by FlorianMinguet** [comment]
> The calculation is also the other way around.
> 
> `fitness = 0.9 * mAP50-95 + 0.1 * mAP50`

I computed again usign this formula and yes the results matches the ones from the fitness function, thanks !

**Comment by FlorianMinguet** [comment]
> How many images do you have? It seems like you have very few validation images.

My dataset is pretty small (68 images), but I wanted to run some tests to progressively see how the model performs depending on the dataset size. Due to the data augmentation only applying to the training set (I used Roboflow to create my dataset), I end up with :
- 130 images for my training set (originally was 42 before data augmentation)
- 14 images for the validation set
- 10 images for the test set

Even though it's likely overfitting, I don't understand how the metrics can show such contrasting results.

**Comment by Y-T-G** [comment]
If you have a small validation set, then the metrics would largely be unreliable because it's too small of a sample size to generalize

**Comment by FlorianMinguet** [comment]
Okay, I'll edit this post when the training will be done, in about an hour

**Comment by glenn-jocher** [comment]
You're welcome! I'm glad everything aligned and worked out for you. If you have any more questions or need further assistance, feel free to ask. Happy experimenting!

**Comment by fazrigading** [comment]
> 1. The metrics highlighted in red at the end of each epoch signify the model performance on the training dataset for that given epoch. This is correct. However, the metrics showcased in blue at the completion of training represent the performance of the 'best' model on the validation dataset. This 'best' model is selected during the course of the entire training based on your criterion of choice (by default, highest mAP).

Simple question:
Why does the number of `Images` every epoch (220 images) is the same as number of validation set images highlighted in blue (at the end of training process) if you said the model evaluate on training set (not validation set) per epoch? It is contradictory, I don't think after every epoch the model did not validating the validation set.

**Comment by Y-T-G** [comment]
The validation is on validation set. mAP, precision and recall are not calculated on training set

**Comment by Anilian** [comment]
@glenn-jocher  I wanted to kindly ask if there is an option to disable model validation on best.pt weights at the end of training. The reason for this is that the final validation step distorts the logged metrics in ClearML, creating a misleading spike in the graphs. This can make it difficult to accurately interpret the training progress.

If it’s possible to adjust this setting, I would greatly appreciate your guidance on how to do so. Thank you for your time and support!

![Image](https://github.com/user-attachments/assets/91603a0b-81a7-4392-bb86-246c55ac2a75)

**Comment by random-guest** [comment]
Please correct me if I am wrong @glenn-jocher 

In ultalytics / utils / metrics.py 

line 817

  def fitness(self):
        """Return model fitness as a weighted combination of metrics."""
        w = [0.0, 0.0, 0.1, 0.9]  # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
        return (np.nan_to_num(np.array(self.mean_results())) * w).sum()

It seems like the best model is a weighted sum of 0.1 * map@0.5 + 0.9*map@0.5-0.95 

Editing these weights would edit how the model will choose the best.pt right?
